src/rvsdg/optimize_direct_jumps.rs

```diff
@@ -99,51 +119,105 @@ impl SimpleCfgFunction {
-            let (empty, target) = self.graph.edge_endpoints(target_edge).unwrap();
-
-            let source_edges: Vec<EdgeIndex> = self
+        // loop over every non-empty block, and look at its parents
+        for block in self.graph.node_indices().collect::<Vec<_>>() {
+            // find parents
+            let parents: Vec<NodeIndex> = self
                 .graph
-                .edges_directed(empty, Direction::Incoming)
-                .map(|edge| edge.id())
+                .edges_directed(block, Direction::Incoming)
+                .map(|edge| edge.source())
                 .collect();
+            let new_parents = parents
+                .iter()
+                .map(|parent| self.get_parent_skipping_empty(*parent))
+                .collect::<HashSet<_>>();
+            let parent_mapping = parents
+                .iter()
+                .map(|parent| {
+                    (
+                        self.graph[*parent].name.to_string(),
+                        self.graph[self.get_parent_skipping_empty(*parent)]
+                            .name
+                            .to_string(),
+                    )
+                })
+                .collect::<HashMap<_, _>>();
+            // if we have fewer parents, phi nodes may get messed up, so avoid the optimization
+            if new_parents.len() == parents.len() {
+                for parent in &parents {
+                    self.collapse_empty_block(*parent);
+                }
 
-            for source_edge in source_edges {
-                let (source, empty_) = self.graph.edge_endpoints(source_edge).unwrap();
-                let weight = self.graph.edge_weight(source_edge).unwrap().clone();
-                assert_eq!(empty, empty_);
-
-                self.graph.remove_edge(source_edge);
-                self.graph.add_edge(source, target, weight);
+                Self::fixup_phis(&mut self.graph[block], &parent_mapping);
             }
         }
 
         self
     }
+
+    fn get_single_in_single_out(&self, parent_block: NodeIndex) -> Option<(EdgeIndex, EdgeIndex)> {
+        if !self.graph[parent_block].instrs.is_empty()
+            || !self.graph[parent_block].footer.is_empty()
+        {
+            return None;
+        }
+
+        let outgoing = self
+            .graph
+            .edges_directed(parent_block, Direction::Outgoing)
+            .collect::<Vec<_>>();
+        let incoming = self
+            .graph
+            .edges_directed(parent_block, Direction::Incoming)
+            .collect::<Vec<_>>();
+        if let ([source_edge], [outgoing]) = (incoming.as_slice(), outgoing.as_slice()) {
+            Some((source_edge.id(), outgoing.id()))
+        } else {
+            None
+        }
+    }
+
+    fn get_parent_skipping_empty(&self, parent_block: NodeIndex) -> NodeIndex {
+        if let Some((source_edge, _)) = self.get_single_in_single_out(parent_block) {
+            self.graph.edge_endpoints(source_edge).unwrap().0
```

Comment:
What if the grandparent is also empty? I think this function should be recursive...

Reply:
Changed the name of the func- it needs to not be recursive for correctness